### PR TITLE
[FIX] mrp_operations_extension: closes #964 When an operation is finished...

### LINKED ISSUE #964: mrp_production_manual_close realiza movimientos de fabricación y consumo
Según parece el módulo mrp_operations de Odoo, interpreta que si se realiza una OF, se dan por realizados todos los movimientos de consumo y producción, incluso cuando la realización de la OF fuese parcial.

Esto está causando un bug grave en el módulo mrp_production_manual_close, ya que al cerrar la OF en vez de cancelar los movimientos pendientes tanto de consumo como de producción, las realiza.
Para solucionarlo, es necesario primero ejecutar la señal de "cancelar" la OF y luego marcarla realizada. De esta forma, se cancela todo lo pendiente y la OF se deja terminada en vez de cancelada.

Además, se va a añadir una pantalla de warning indicando lo que hace el  botón. Algo así como:
"Esta acción cancelará todos los movimientos pendientes tanto de consumo como de producción y marcará la OF como realizada y no se podrá deshacer. ¿Está seguro?"

### PR DESCRIPTION
**Pull request by Daniel-CA**
When an operation is finished it should not produce with the remaining raw materials.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I don't understand why you cancel moves on action_done. This way, you were never able to consume and produce.

File: mrp_operations_extension/models/mrp_production.py [hunk 1]
```diff
@@ -157,3 +157,14 @@ def action_start_working(self):
         if self.production_id.state in ('confirmed', 'ready'):
             self.production_id.state = 'in_production'
         return super(MrpProductionWorkcenterLine, self).action_start_working()
+
+    def action_done(self):
+        for move_line in self.move_lines:
```
Comment (pedrobaeza) [inline mrp_operations_extension/models/mrp_production.py]: As exercise, you should start thinking with new API possibilities: 
`self.move_line.filtered(lambda x: x not in ('cancel', 'done')).action_cancel()`

Comment (anajuaristi) [inline mrp_operations_extension/models/mrp_production.py]: We detected a bug on operations extension. If one operation is finished all pending moves have to be canceled. Allowing to consume or to produce after ending the operation was causing big troubles on standar MO workflow and it's not conceptually correct.
This fix is reparing the issue

**Comment by Daniel-CA** [comment]
Done.

**Comment by Daniel-CA** [comment]
Anything more to change here?

**Comment by pedrobaeza** [comment]
I still don't see this movement, so I can't approve it. Go ahead if you want, but I'm not sure this will have side effects.

**Comment by Daniel-CA** [comment]
I spoke with Ana about the side effects that is going to have to change the action_done() function for WO in Odoo. Now in Odoo when you finish a WO you consume/produced all the material you have, with this focus when you finish a WO all the "remaining" material for consume/produce is cancelled. Neither I am about if this is going to have side effects, certainly we are changing how WO works.

**Comment by pedrobaeza** [comment]
You should warn it then, but still doesn't convince to change the workflow this way. It should do it on cancel, not on done action.

**Comment by anajuaristi** [comment]
I will put an example about this:
Supose I need to produce 10 PF and consume 20RM

I start the order and I produce 9 PF and really consumed 19RM (notice it's
19, not 18)
So it remains 1 PF stock_move from production to stock and anoter 1 RM
stock_move from stock to production

At this point without the fix, if I close the operation, Odoo will mark
DONE both remaining moves so it's breaking stock, since it's not true.
If we cancel the order... it's also not true because we produced a part, so
it's done, not cancelled. If we cancel, we have not clue about really
cancelled orders and the ones that has been partially done.

The correct way is cancelling remaining moves, so stock is correct and
making the order done, because it has been finished.

For me it's RTM :+1:

**Comment by anajuaristi** [comment]
I agree with Pedro on "You should warn it then"

I would put a warning on pushing the finish button with a text about what it's going to happen

**Comment by pedrobaeza** [comment]
I consider nearer the cancel semantic than done semantic, although the ideal is to add a new state called "Closed" to reflect this.

**Comment by anajuaristi** [comment]
Yeahh... it's conceptually complex to define de right one but... ¿Is it necesary adding a new status on operations or MO's?

One collateral effect we detected if we use cancel on OF is that all operations are also cancelled. 
Supose, you consume everything and finish all "consume operations" without any change but... in last step you only produce 1 PF less than expected(it's broken or something) so... if you push cancel OF will mark cancelled also all "done" operations. I know that it could be controlled by code but It was becaming too complex... so... I think, this way with a warning we can avoid the new status "closed" and things are finished correctly at this point but... 

On the other side,  I see also correct to add a new "closed" status (In MO and Operations), having:
 "done" as "everything went as expected" 
"closed" as "we did something with it but we did not made it as expected". Just added more or less materials, or changed quantities or changed anything. 

What do you think? Could it be usefull having MO's and Operations in 2 different statuses?

**Comment by pedrobaeza** [comment]
I think so, so we should add this state. It's very easy and inheritable with new API, so no problem in that technical part. You only have to check the workflows for getting all correct transitions.

**Comment by pedrobaeza** [comment]
Any news?

**Comment by anajuaristi** [comment]
I was not sure of adding a new state in MO, so I tried to find something that covers both options... 

The answer is very easy... when someone pushes the button, if there is movements in state <> from "cancel", "done", we should show a warning screen asking:

"There is still some pending moves on MO, do you want to "make them done", "cancel them", "discard"

make them done: Realizes all movements and closes the order
cancel them: cancels all movements and closes the order
discard: nothing is done, order remains open.

@pedrobaeza Opinion? I think this way is too much easy and covers perfectly the needed functionality

---

On the other side, it would be needed to add exactly same functionality on buton "finish" on work orders. 
Actually, the system allows finishing an operation and pending movements remain  pending. I think it's not correct and we should control same way defined for full MO.

Again, opinion?

**Comment by pedrobaeza** [comment]
OK, this approach seems correct to me

**Comment by Daniel-CA** [comment]
Cambios realizados, a ver como lo veis.

**Comment by pedrobaeza** [comment]
Revisa el error de Travis.

@anajuaristi, comprueba funcionalmente.

**Comment by Daniel-CA** [comment]
El error del Travis según pone es en el purchase_product_variants_types, que no encuentra el custom_value. :confused:

**Comment by pedrobaeza** [comment]
OK, reviso yo ahora eso

**Comment by pedrobaeza** [comment]
Haz un rebase, que te debe salir ya bien

**Comment by anajuaristi** [comment]
@Daniel-CA @pedrobaeza sigue en rojo. Me avisais cuando se pueda testear? 
Gracias

File: mrp_operations_extension/wizard/workcerter_line_finish_view.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            <field name="name">finish.wo.form.view</field>
+            <field name="model">workcenter.line.finish</field>
+            <field name="arch" type="xml">
+                <form string="Finish Work Order">
+                    <separator
+                        string="There are still some pending moves on WO"
+                        colspan="4" />
+                    <footer>
+                        <button class="oe_highlight" name="make_them_done"
+                            string="Done all movements" type="object" />
```
Comment (oihane) [inline mrp_operations_extension/wizard/workcerter_line_finish_view.xml]: @Daniel-CA Won't it be "Do all movements"?

**Comment by pedrobaeza** [comment]
Es tema de pylint. @Daniel-CA, revisa.

@anajuaristi, puedes probar.

**Comment by Daniel-CA** [comment]
Ahora dan fallos los test del quality_control_mrp, tiene que ver con el mrp_hook?

**Comment by pedrobaeza** [comment]
No, es porque el módulo delivery pone un campo obligatorio (weight_uom_id), pero todavía los default no están activos, creando un error ficticio. Es difícil de corregir, la verdad...

**Comment by Daniel-CA** [comment]
Me comenta @oihane que se ha movido ya de rama el mrp_operations_extension, no se si se han tenido en cuenta las modificaciones que meti aquí para el funcionamiento de las OT y el control que de cantidad a producir que añadí para arreglar el bug que se generaba al cerrar una OF con el botón del módulo mrp_production_manual_close

**Comment by pedrobaeza** [comment]
No, no se han tenido en cuenta. Habría que hacer un PR con el nuevo código adaptado.

**Pull request closed without merging**